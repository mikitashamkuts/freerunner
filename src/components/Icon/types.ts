import {iconColorList, iconList} from '@src/design';
import {CustomComponentProps} from '@src/types';

export interface Props extends CustomComponentProps {
  name: keyof typeof iconList;
  color: keyof typeof iconColorList;
}
