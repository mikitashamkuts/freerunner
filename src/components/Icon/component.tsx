import React, {FC, memo} from 'react';
import {StyleSheet, View} from 'react-native';
import {Path, Svg} from 'react-native-svg';

import {iconColorList, iconList} from '@src/design';

import {whyDidItRenderConfig} from '../../../debug';

import {Props, iconSettings} from '.';

/**
 * Icon component renders an SVG icon based on the provided name and color.
 *
 * @param {IconProps} props - The props for the component.
 * @returns {JSX.Element | null} The Icon component or null if no name is provided.
 */
const Icon: FC<Props> = ({name, color, containerStyle}) => {
  if (!name) {
    return null;
  }

  const iconConfig = iconSettings[iconList[name]];
  const iconHeight = iconConfig.settings.height;
  const iconWidth = iconConfig.settings.width;

  const styles = StyleSheet.create({
    container: {
      overflow: 'hidden',
      height: iconHeight,
      width: iconWidth,
    },
    iconContainer: {
      left: iconConfig.settings.customShift?.left,
      top: iconConfig.settings.customShift?.top,
    },
  });

  return (
    <View style={[styles.container, containerStyle]}>
      <Svg
        style={styles.iconContainer}
        height={iconHeight}
        width={iconWidth}
        viewBox={iconConfig.settings.viewBox}
        fill="none">
        <Path
          d={iconConfig.paths.d1}
          fill={color ? iconColorList[color] : iconConfig.settings.color}
        />
        <Path
          fillRule="evenodd"
          d={iconConfig.paths.d2}
          fill={color ? iconColorList[color] : iconConfig.settings.color}
        />
      </Svg>
    </View>
  );
};

Icon.whyDidYouRender = whyDidItRenderConfig.UIComponentDebugActive;

export default memo(Icon);
