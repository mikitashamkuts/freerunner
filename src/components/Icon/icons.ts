import {iconColorList, iconList} from '../../design';
import {IconSettingsType} from '../../types';

export const iconSettings: IconSettingsType = {
  [iconList.Arrow]: {
    paths: {
      d1: 'M0.289062 10.2441C0.296224 10.0221 0.339193 9.81803 0.417969 9.63184C0.503906 9.44564 0.632812 9.2666 0.804688 9.09473L9.07617 1.00586C9.34831 0.733724 9.68132 0.597656 10.0752 0.597656C10.3473 0.597656 10.5908 0.662109 10.8057 0.791016C11.0277 0.919922 11.2031 1.09538 11.332 1.31738C11.4681 1.53223 11.5361 1.77214 11.5361 2.03711C11.5361 2.43815 11.3857 2.78906 11.085 3.08984L3.72656 10.2441L11.085 17.3984C11.3857 17.7064 11.5361 18.0573 11.5361 18.4512C11.5361 18.7233 11.4681 18.9668 11.332 19.1816C11.2031 19.3965 11.0277 19.5684 10.8057 19.6973C10.5908 19.8262 10.3473 19.8906 10.0752 19.8906C9.68132 19.8906 9.34831 19.7546 9.07617 19.4824L0.804688 11.3936C0.625651 11.2217 0.493164 11.0426 0.407227 10.8564C0.328451 10.6631 0.289062 10.459 0.289062 10.2441Z',
    },
    settings: {
      color: iconColorList.Action,
      width: 12,
      height: 20,
      viewBox: '0 0 12 20',
    },
  },
  [iconList.Plus]: {
    paths: {
      d1: 'M0.637695 9.24414C0.637695 8.97917 0.730794 8.75358 0.916992 8.56738C1.11035 8.37402 1.33594 8.27734 1.59375 8.27734H8.54395V1.33789C8.54395 1.08008 8.63704 0.858073 8.82324 0.671875C9.00944 0.478516 9.23503 0.381836 9.5 0.381836C9.76497 0.381836 9.99056 0.478516 10.1768 0.671875C10.3701 0.858073 10.4668 1.08008 10.4668 1.33789V8.27734H17.4062C17.6641 8.27734 17.8861 8.37402 18.0723 8.56738C18.2656 8.75358 18.3623 8.97917 18.3623 9.24414C18.3623 9.50911 18.2656 9.7347 18.0723 9.9209C17.8861 10.1071 17.6641 10.2002 17.4062 10.2002H10.4668V17.1504C10.4668 17.4082 10.3701 17.6302 10.1768 17.8164C9.99056 18.0098 9.76497 18.1064 9.5 18.1064C9.23503 18.1064 9.00944 18.0098 8.82324 17.8164C8.63704 17.6302 8.54395 17.4082 8.54395 17.1504V10.2002H1.59375C1.33594 10.2002 1.11035 10.1071 0.916992 9.9209C0.730794 9.7347 0.637695 9.50911 0.637695 9.24414Z',
    },
    settings: {
      color: iconColorList.Action,
      width: 19,
      height: 19,
      viewBox: '0 0 19 19',
    },
  },
};
