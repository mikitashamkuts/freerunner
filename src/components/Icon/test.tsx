/**
 * @format
 */

import 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
// Note: import explicitly to use the types shipped with jest.
import {it} from '@jest/globals';

import Icon from './component';

describe('Icon Component', () => {
  it('Renders with a given icon name and default color and matches previous snapshot: ', () => {
    const tree = renderer.create(<Icon color="Default" name="BagXS" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with a given icon name and custom color and matches previous snapshot: ', () => {
    const tree = renderer.create(<Icon name="Arrow" color="Inversive" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with additional container style and matches previous snapshot: ', () => {
    const customStyle = {margin: 10};
    const tree = renderer
      .create(<Icon color="Default" name="CheckBig" containerStyle={customStyle} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Returns null when name is not provided and matches previous snapshot: ', () => {
    const tree = renderer.create(<Icon />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
