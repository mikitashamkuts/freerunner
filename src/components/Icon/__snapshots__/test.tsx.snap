// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Component Renders with a given icon name and custom color and matches previous snapshot:  1`] = `
<View
  style={
    [
      {
        "height": 24,
        "overflow": "hidden",
        "width": 24,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="24"
    bbWidth="24"
    fill="none"
    focusable={false}
    height={24}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "left": undefined,
          "top": undefined,
        },
        {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={24}
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
        ]
      }
    >
      <RNSVGPath
        d="M19 11H7.82998L12.71 6.12C13.1 5.73 13.1 5.09 12.71 4.7C12.32 4.31 11.69 4.31 11.3 4.7L4.70998 11.29C4.31998 11.68 4.31998 12.31 4.70998 12.7L11.3 19.29C11.69 19.68 12.32 19.68 12.71 19.29C13.1 18.9 13.1 18.27 12.71 17.88L7.82998 13H19C19.55 13 20 12.55 20 12C20 11.45 19.55 11 19 11Z"
        fill={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`Icon Component Renders with a given icon name and default color and matches previous snapshot:  1`] = `
<View
  style={
    [
      {
        "height": 16,
        "overflow": "hidden",
        "width": 16,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="16"
    bbWidth="16"
    fill="none"
    focusable={false}
    height={16}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "left": -0.5,
          "top": undefined,
        },
        {
          "flex": 0,
          "height": 16,
          "width": 16,
        },
      ]
    }
    vbHeight={16}
    vbWidth={16}
    width={16}
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
        ]
      }
    >
      <RNSVGPath
        d="M10.4406 2.25702C10.8351 2.72174 11.0196 3.36739 11.0168 4.19615L11.0115 4.42672L11.0004 4.60504L11.0771 4.60604C11.8943 4.60604 12.6149 5.10211 12.922 5.84972L12.9778 6.00253L13.022 6.1616L13.0513 6.32007L13.9802 12.8956C14.1271 13.9357 13.4421 14.9015 12.4162 15.1228L12.2673 15.1492L12.1282 15.1627L12.006 15.1667H3.99349C2.9424 15.1667 2.08127 14.3538 2.00543 13.3076L2 13.1566L2.00612 13.0171L2.01927 12.8956L2.94815 6.32007C3.07 5.45748 3.73571 4.78254 4.59747 4.63442L4.76127 4.6129L4.92206 4.60604H5.0499L5.04026 4.435C5.00898 3.66607 5.12696 3.03907 5.40946 2.55341L5.50974 2.39676L5.61158 2.26343C6.82365 0.801595 9.20605 0.802693 10.4406 2.25702ZM9.24286 6.1991H6.80843L6.94007 6.6715L7.03273 6.9782L7.11054 7.02859C7.37809 7.22373 7.5575 7.52537 7.59406 7.86458L7.60097 7.99341C7.60097 8.65369 7.06537 9.18895 6.40469 9.18895C5.744 9.18895 5.20841 8.65369 5.20841 7.99341C5.20841 7.68081 5.32961 7.39069 5.53319 7.17429L5.56031 7.14657L5.47735 6.86115L5.37844 6.49351L5.3061 6.1991L4.92237 6.2001C4.76746 6.2001 4.62941 6.28946 4.5683 6.40505L4.54443 6.46474L4.52753 6.5429L3.59865 13.1184C3.59604 13.1368 3.59473 13.1555 3.59473 13.1741C3.59473 13.3503 3.71003 13.5019 3.85042 13.5517L3.91185 13.5666L3.99349 13.5726L12.0339 13.5716L12.0617 13.5687C12.2327 13.5446 12.3649 13.4142 12.3982 13.2718L12.4062 13.2L12.4008 13.1184L11.4719 6.5429C11.4503 6.38961 11.3424 6.26549 11.2194 6.22117L11.1569 6.2059L11.0766 6.2001L10.7432 6.1991L10.6849 6.44851L10.6262 6.67604L10.48 7.1924L10.5364 7.25599C10.6709 7.42736 10.7581 7.63481 10.7834 7.85786L10.791 7.99341C10.791 8.65369 10.2555 9.18895 9.59477 9.18895C8.93408 9.18895 8.39849 8.65369 8.39849 7.99341C8.39849 7.56167 8.62916 7.17686 8.98191 6.96639L9.03451 6.93834L9.12474 6.63503L9.24286 6.1991ZM6.74231 3.17307C6.55899 3.39416 6.46897 3.81456 6.50113 4.42754L6.51335 4.60504H9.53695L9.55029 4.42153C9.58094 3.86641 9.51354 3.47303 9.37095 3.24818L9.32067 3.17948C8.67152 2.4148 7.3723 2.41326 6.74231 3.17307Z"
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`Icon Component Renders with additional container style and matches previous snapshot:  1`] = `
<View
  style={
    [
      {
        "height": 117,
        "overflow": "hidden",
        "width": 117,
      },
      {
        "margin": 10,
      },
    ]
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="117"
    bbWidth="117"
    fill="none"
    focusable={false}
    height={117}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "left": undefined,
          "top": undefined,
        },
        {
          "flex": 0,
          "height": 117,
          "width": 117,
        },
      ]
    }
    vbHeight={117}
    vbWidth={117}
    width={117}
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
        ]
      }
    >
      <RNSVGPath
        d="M58.5 0.375C90.5493 0.375 116.625 26.4507 116.625 58.5C116.625 90.5493 90.5493 116.625 58.5 116.625C26.4507 116.625 0.375 90.5493 0.375 58.5C0.375 26.4507 26.4507 0.375 58.5 0.375ZM58.5 6.18276C29.6508 6.18276 6.18276 29.6508 6.18276 58.5C6.18276 87.3492 29.6508 110.817 58.5 110.817C87.3492 110.817 110.817 87.3492 110.817 58.5C110.817 29.6508 87.3492 6.18276 58.5 6.18276ZM78.4845 39.9165C79.4107 39.8525 80.3019 40.1528 80.9994 40.7593C81.6969 41.3638 82.1193 42.2066 82.1832 43.1288C82.2491 44.0511 81.9488 44.9462 81.3385 45.6476L54.6959 76.7251C54.041 77.4904 53.1052 77.9205 52.0667 77.936C51.096 77.936 50.1641 77.5253 49.5092 76.8084L36.5977 62.6821C35.3074 61.2735 35.4062 59.0764 36.8167 57.788C37.4541 57.2028 38.2833 56.8812 39.1513 56.8812C40.124 56.8812 41.0559 57.292 41.7088 58.0069L51.9834 69.2483L76.115 41.1003C76.7195 40.4028 77.5604 39.9823 78.4845 39.9165Z"
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        d="M58.5 0.375C90.5493 0.375 116.625 26.4507 116.625 58.5C116.625 90.5493 90.5493 116.625 58.5 116.625C26.4507 116.625 0.375 90.5493 0.375 58.5C0.375 26.4507 26.4507 0.375 58.5 0.375ZM58.5 6.18276C29.6508 6.18276 6.18276 29.6508 6.18276 58.5C6.18276 87.3492 29.6508 110.817 58.5 110.817C87.3492 110.817 110.817 87.3492 110.817 58.5C110.817 29.6508 87.3492 6.18276 58.5 6.18276ZM78.4845 39.9165C79.4107 39.8525 80.3019 40.1528 80.9994 40.7593C81.6969 41.3638 82.1193 42.2066 82.1832 43.1288C82.2491 44.0511 81.9488 44.9462 81.3385 45.6476L54.6959 76.7251C54.041 77.4904 53.1052 77.9205 52.0667 77.936C51.096 77.936 50.1641 77.5253 49.5092 76.8084L36.5977 62.6821C35.3074 61.2735 35.4062 59.0764 36.8167 57.788C37.4541 57.2028 38.2833 56.8812 39.1513 56.8812C40.124 56.8812 41.0559 57.292 41.7088 58.0069L51.9834 69.2483L76.115 41.1003C76.7195 40.4028 77.5604 39.9823 78.4845 39.9165Z"
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
      <RNSVGPath
        fill={
          {
            "payload": 4280032284,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`Icon Component Returns null when name is not provided and matches previous snapshot:  1`] = `null`;
